samples/nrf9160/modem_shell/src/sock/sock.c

```diff
@@ -0,0 +1,1320 @@
+		shell_error(
+			shell_global,
+			"Port (%d) must be bigger than 0 and smaller than 65536",
+			port);
+		return -EINVAL;
+	}
+
+	/* Validate bind port. Zero means that binding is not done. */
+	if (bind_port > 65535) {
+		shell_error(
+			shell_global,
+			"Bind port (%d) must be smaller than 65536",
+			bind_port);
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static int sock_protocol_resolve(
+	int family,
+	int type,
+	bool secure)
+{
+	int proto = 0;
+
+	/* Map type to protocol.
+	 * If proto is set to zero for stream or dgram socket to let lower stack select it,
+	 * socket creation fails with errno=43 (PROTONOSUPPORT). Hence we need to set protocol.
+	 */
+	if (type == SOCK_STREAM) {
+		proto = IPPROTO_TCP;
+	} else if (type == SOCK_DGRAM) {
+		proto = IPPROTO_UDP;
+	}
+	/* Proto is left to zero for raw sockets */
+
+	if (secure) {
+		if (type == SOCK_STREAM) {
+			proto = IPPROTO_TLS_1_2;
+		} else if (type == SOCK_DGRAM) {
+			proto = IPPROTO_DTLS_1_2;
+		}
+	}
+
+	return proto;
+}
+
+static int sock_getaddrinfo(
+	struct sock_info_t *socket_info,
+	int family,
+	int type,
+	char *address,
+	int port)
+{
+	int err;
+
+	if ((address == NULL) || (strlen(address) == 0)) {
+		if (type != SOCK_RAW) {
+			shell_error(shell_global, "Address not given");
+			return -EINVAL;
+		}
+	} else {
+		struct addrinfo hints = {
+			.ai_family = family,
+			.ai_socktype = type,
+		};
+		err = getaddrinfo(address, NULL, &hints, &socket_info->addrinfo);
+		if (err) {
+			shell_error(
+				shell_global,
+				"getaddrinfo() failed, err %d errno %d",
+				err,
+				errno);
+			err = errno;
+			return err;
+		}
+
+		/* Set port to address info */
+		if (family == AF_INET) {
+			((struct sockaddr_in *)socket_info->addrinfo->ai_addr)->sin_port =
+				htons(port);
+		} else if (family == AF_INET6) {
+			((struct sockaddr_in6 *)socket_info->addrinfo->ai_addr)->sin6_port =
+				htons(port);
+		} else {
+			assert(0);
+		}
+	}
+	return 0;
+}
+
+static int sock_set_tls_options(
+	int fd,
+	int sec_tag,
+	bool session_cache,
+	int peer_verify,
+	char *peer_hostname)
+{
+	int err;
+	sec_tag_t sec_tag_list[] = { sec_tag };
+	uint8_t cache;
+	uint32_t verify;
+
+	/* Security tag */
+	err = setsockopt(fd, SOL_TLS, TLS_SEC_TAG_LIST, sec_tag_list,
+			 sizeof(sec_tag_t) * ARRAY_SIZE(sec_tag_list));
+	if (err) {
+		shell_error(shell_global, "Unable to set security tag, errno %d", errno);
+		err = errno;
+		return -EINVAL;
+	}
+
+	/* Session cache */
+	if (session_cache) {
+		cache = TLS_SESSION_CACHE_ENABLED;
+	} else {
+		cache = TLS_SESSION_CACHE_DISABLED;
+	}
+	err = setsockopt(fd, SOL_TLS, TLS_SESSION_CACHE, &cache, sizeof(cache));
+	if (err) {
+		shell_error(shell_global, "Unable to set session cache, errno %d", errno);
+		return errno;
+	}
+
+	/* Peer verify */
+	switch (peer_verify) {
+	case 0:
+		verify = TLS_PEER_VERIFY_NONE;
+		break;
+	case 2:
+		verify = TLS_PEER_VERIFY_REQUIRED;
+		break;
+	case 1:
+	default:
+		verify = TLS_PEER_VERIFY_OPTIONAL;
```

Comment:
Why's optional the default behaviour? It'll cause the handshake to proceed even if the certificate verification fails.

Reply:
Optional is the modem default, that's why the same is used also here. I think it makes sense to use the same default values which the modem uses, even though from security point of view required would be better.

Reply:
Ok, noted. I'm a bit surprised though to hear that the modem FW defaults to OPTIONAL...

Reply:
Sorry, I don't know where I got the impression that the default was "optional". I checked the modem lib code and the default is "required". And I think that modem doesn't actually have a default, but will use the value given in the RPC API. I will change the default value to "required", that's definitely better.